securedrop_client/gui/widgets.py

```diff
@@ -2368,37 +2374,65 @@ def clear_conversation(self):
                 child.widget().deleteLater()
 
     def update_conversation(self, collection: list) -> None:
-        # clear all old items
-        self.clear_conversation()
+        """
+        Given a list of conversation items that reflect the new state of the
+        conversation, this method does two things:
+
+        * Checks if the conversation item already exists in the conversation.
+          If so, it checks that it's still in the same position. If it isn't,
+          the item is removed from its current position and re-added at the
+          new position. Then the index meta-data on the widget is updated to
+          reflect this change.
+        * If the item is a new item, this is created (as before) and inserted
+          into the conversation at the correct index.
+
+        Things to note, speech bubbles and files have an index attribute which
+        defines where they currently are. This is the attribute that's checked
+        when the new conversation state (i.e. the collection argument) is
+        passed into this method in case of a mismatch between where the widget
+        has been and now is in terms of its index in the conversation.
+        """
         self.controller.session.refresh(self.source)
-        # add new items
-        for conversation_item in collection:
-            if isinstance(conversation_item, Message):
-                self.add_message(conversation_item)
-            elif isinstance(conversation_item, (DraftReply, Reply)):
-                self.add_reply(conversation_item)
+        for index, conversation_item in enumerate(collection):
+            item_widget = self.current_messages.get(conversation_item.uuid)
+            if item_widget:
+                # check an already displayed item.
+                if item_widget.index != index:
+                    # The existing widget is out of order, remove / re-add it
+                    # and update index details.
+                    self.conversation_layout.removeWidget(item_widget)
+                    item_widget.index = index
+                    self.conversation_layout.insertWidget(index, item_widget)
+                # TODO: Check if text in item has changed, then update the
```

Comment:
we would want to do something similar to ReplyWidget for MessageWidget (we connect the success and failure signals to each widget so that we can update the speech bubble bar and error indicator)

so we need to connect the message download signals to message widget handlers so we can update the text inline in a similar fashion

Reply:
Ack... I'll look into all this signal plumbing. ;-)

If this PR is step 1, your suggesting is part of step 2, and before we merge we need to make sure we address step 3.

1. Make it work.
2. Make it work properly.
3. Make it work properly but with tests and documentation. 

:-)

I imagine we'll be spending some time to make sure step 2 is comprehensive and complete. All suggestions welcome.